File: packages/pstack/src/subdomains.ts
/**
 * Wildcard subdomain routing for a per-PR surface.
 *
 * THE GOAL. Given profile `backend` on stack `pr-123`, make anything under
 * `backend-pr-123.<domain>` reach the same place `backend-pr-123.<domain>` reaches —
 * `api.backend-pr-123.<domain>`, `tenant-a.backend-pr-123.<domain>`, and (opt-in) deeper still.
 * Apps that route by subdomain — per-tenant hosts, per-branch previews inside a preview — need it,
 * and hard-coding one host per tenant into a per-PR compose file does not scale.
 *
 * HOW IT PLUGS IN, AND WHY THIS WAY. pstack does not own your per-PR Traefik labels; your compose
 * file does (see docs/control-plane.md § Hostnames). So this module does not write labels, generate
 * an overlay, or touch Traefik's dynamic directory. It computes the one part that is genuinely
 * fiddly — a correctly escaped, correctly anchored Go regexp — and exposes it as an environment
 * variable that `docker compose` interpolates into a label you write:
 *
 *     labels:
 *       - traefik.http.routers.backend-wild.rule=${PSTACK_WILD_BACKEND}
 *       - traefik.http.routers.backend-wild.priority=2
 *       - traefik.http.routers.backend-wild.service=backend
 *
 * Three alternatives were rejected:
 *
 *   - **Merging labels via a generated overlay.** Compose's merge semantics for list-form `labels`
 *     decide whether your own routers survive that merge, and getting it wrong deletes them
 *     silently. Not a risk worth taking for a convenience feature.
 *   - **Writing Traefik file-provider config.** Would need the API container mounted into
 *     `traefik-dynamic/` read-write, which means every existing host must re-run `pstack init` —
 *     and one malformed file in that directory takes down routing for *everything* on the box, by
 *     Traefik's own documented behaviour.
 *   - **A label-carrying sidecar container per PR.** A whole container to hold three strings.
 *
 * The env-var seam needs none of that: no new mount, no re-`init`, no capability added to the API,
 * and nothing pstack writes can break another deployment's routing.
 *
 * "UNLESS IT'S HARDCODED" IS FREE. Traefik's default router priority is the rule's length, so an
 * exact `Host(`admin.backend-pr-123.example.com`)` scores in the dozens and beats a wildcard router
 * pinned at `priority=2`. Pin it explicitly anyway: a *generated* number that happens to be lower
 * is not a guarantee, and 2 also clears the `preview-fallback` catch-all at priority 1.
 *
 * ── THE CEILINGS. Read these before promising anyone arbitrary depth. ────────────────────────────
 *
 * Routing is the easy third of the problem. The other two are outside pstack:
 *
 *   1. **DNS.** A wildcard record matches exactly ONE label. `*.example.com` answers for
 *      `a.example.com` and NOT for `a.b.example.com`, and `*.*.example.com` is not a valid record.
 *      So `depth: any` needs a resolver that answers for arbitrary depth — a self-hosted
 *      authoritative server, or a provider with a non-standard extension. Most managed DNS cannot.
 *   2. **TLS.** A certificate wildcard also matches exactly one label, so
 *      `*.<domain>` does NOT cover `x.backend-pr-123.<domain>`. Serving that over HTTPS needs a
 *      certificate for `*.backend-pr-123.<domain>` — which means **DNS-01** (HTTP-01 cannot issue
 *      wildcards at all) and **one certificate per PR**, against Let's Encrypt's ~50
 *      certificates/registered-domain/week. And **no certificate can cover `depth: any`**:
 *      `*.*.host` is not a legal SAN. Arbitrary depth is therefore HTTP-only, permanently.
 *
 * Hence `depth` defaults to `one`: the depth DNS and TLS can actually deliver. `any` is available
 * because the request was explicit about wanting it, and is honest about being routing-only.
 */

import { SpecError } from './errors.ts';

/** How many labels the wildcard spans. `one` is what DNS and TLS can cover; `any` is HTTP-only. */
export type SubdomainDepth = 'one' | 'any';

export type SubdomainRoute = {
  /** The compose profile this belongs to. Also the default hostname prefix. */
  profile: string;
  /**
   * The host the wildcard sits under, fully resolved —
   * e.g. `backend-pr-123.preview.example.com`.
   */
  host: string;
  depth: SubdomainDepth;
  /** The env var carrying the rule, e.g. `PSTACK_WILD_BACKEND`. */
  varName: string;
  /** The Traefik rule, ready to paste into a router label. */
  rule: string;
};

/**
 * Escape a hostname for embedding in a Go regexp.
 *
 * Only `.` needs it in practice — a hostname cannot contain the other metacharacters — but escaping
 * the full set costs nothing and means a future relaxation of what a host may contain cannot turn
 * this into a rule that matches more than intended. An unescaped dot is the classic version of that
 * bug: `backend-pr-1.example.com` would also match `backend-pr-1Xexample.com`.
 */
export function escapeHostRegexp(host: string): string {
  return host.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * The Traefik rule for one wildcard host.
 *
 * `depth: one`  →  exactly one extra label: `api.<host>` yes, `a.b.<host>` no.
 * `depth: any`  →  one or more: both of the above.
 *
 * Anchored at both ends, and the label class excludes `.` so `one` really means one. The bare host
 * itself is deliberately NOT matched — that is your existing exact-host router's job, and matching
 * it here would put two routers on the same hostname for no reason.
 */
export function wildcardRule(host: string, depth: SubdomainDepth = 'one'): string {
  const h = escapeHostRegexp(host);
  const label = '[a-z0-9]([a-z0-9-]*[a-z0-9])?';
  const prefix = depth === 'any' ? `(${label}\\.)+` : `${label}\\.`;
  // Traefik v3 rules quote their argument with BACKTICKS, not quotes.
  return `HostRegexp(\`^${prefix}${h}$\`)`;
}

/**
 * `backend` → `PSTACK_WILD_BACKEND`.
 *
 * Dashes become underscores because an environment variable name cannot hold a dash — which makes
 * `admin-api` and `admin_api` collide, so the caller checks for that (below) rather than letting one
 * profile's rule silently overwrite another's.
 */
export function subdomainVarName(profile: string): string {
  return `PSTACK_WILD_${profile.toUpperCase().replace(/-/g, '_')}`;
}

/** Parsed form of one `compose.subdomains` entry, before the host is resolved. */
export type SubdomainConfig = {
  profile: string;
  depth: SubdomainDepth;
  /** Explicit host, if the spec gave one. Otherwise derived as `<profile>-<stack>.<domain>`. */
  host?: string;
};

/**
 * Parse `compose.subdomains`.
 *
 * Two accepted shapes, because the short one covers the common case and the long one is needed the
 * moment a hostname does not follow the convention:
 *
 *     subdomains: [backend, frontend]                  # depth: one, host derived
 *     subdomains:
 *       backend: any                                   # just the depth
 *       frontend: { depth: one, host: app-${STACK}.${DOMAIN} }
 */
export function parseSubdomains(
  raw: unknown,
  where: string,
  interp: (v: string, at: string) => string,
): SubdomainConfig[] {
  if (raw === undefined) return [];

  const depthOf = (v: unknown, at: string): SubdomainDepth => {
    if (v === undefined) return 'one';
    if (v !== 'one' && v !== 'any') {
      throw new SpecError(
        `${at} must be "one" (a single label — what DNS and TLS can cover) or "any" ` +
          `(any depth, routing only: no certificate can cover it). Got ${JSON.stringify(v)}.`,
      );
    }
    return v;
  };

  if (Array.isArray(raw)) {
    return raw.map((p, i) => {
      if (typeof p !== 'string') {
        throw new SpecError(`${where}[${i}] must be a profile name`);
      }
      return { profile: interp(p, `${where}[${i}]`), depth: 'one' as const };
    });
  }

  if (typeof raw !== 'object' || raw === null) {
    throw new SpecError(`${where} must be a list of profile names or a mapping of profile to options`);
  }

  return Object.entries(raw as Record<string, unknown>).map(([profile, v]) => {
    const at = `${where}.${profile}`;
    if (typeof v === 'string' || v === undefined || v === null) {
      return { profile, depth: depthOf(v ?? undefined, at) };
    }
    if (typeof v !== 'object' || Array.isArray(v)) {
      throw new SpecError(`${at} must be a depth ("one"/"any") or a mapping of { depth?, host? }`);
    }
    const m = v as Record<string, unknown>;
    for (const k of Object.keys(m)) {
      if (k !== 'depth' && k !== 'host') {
        throw new SpecError(`${at}.${k} is not a known option — expected \`depth\` or \`host\``);
      }
    }
    const host = m.host === undefined ? undefined : interp(String(m.host), `${at}.host`);
    return { profile, depth: depthOf(m.depth, `${at}.depth`), host };
  });
}

/**
 * Resolve every configured subdomain into a concrete route.
 *
 * `domain` is required and comes from the spec's own `env: DOMAIN:` — the same variable a per-PR
 * compose file already interpolates for its exact-host routers, so there is nothing new to declare
 * in the usual case. It is a HARD error when missing rather than a rule anchored on an empty
 * suffix: `HostRegexp(^[a-z0-9-]+\.$)` matches nothing at all, and a router that silently never
 * fires is worse than a spec that refuses to parse.
 */
export function resolveSubdomains(args: {
  configs: SubdomainConfig[];
  stack: string;
  profiles: string[];
  domain?: string;
}): SubdomainRoute[] {
  const { configs, stack, profiles, domain } = args;
  if (configs.length === 0) return [];

  if (!domain) {
    throw new SpecError(
      '`compose.subdomains` needs a domain to anchor its rules to. Declare it in the spec ' +
        '(`env:\n  DOMAIN: preview.example.com`) or give each entry an explicit `host:`. ' +
        'Without it the generated rule would match nothing.',
    );
  }

  const seenVar = new Map<string, string>();
  return configs.map((c) => {
    // A subdomain rule for a profile that is never brought up is dead config, and the likeliest
    // cause is a typo in one of the two lists — so say which names ARE available.
    if (!profiles.includes(c.profile)) {
      throw new SpecError(
        `compose.subdomains names profile "${c.profile}", which is not in compose.profiles ` +
          `(${profiles.length ? profiles.join(', ') : 'empty'}). Its router would never match a ` +
          `running service.`,
      );
    }
    const varName = subdomainVarName(c.profile);
    const clash = seenVar.get(varName);
    if (clash !== undefined) {
      throw new SpecError(
        `profiles "${clash}" and "${c.profile}" both map to ${varName} (a dash and an underscore ` +
          `are the same character in an environment variable name), so one rule would overwrite ` +
          `the other. Rename one profile.`,
      );
    }
    seenVar.set(varName, c.profile);

    const host = c.host ?? `${c.profile}-${stack}.${domain}`;
    return { profile: c.profile, host, depth: c.depth, varName, rule: wildcardRule(host, c.depth) };
  });
}

/**
 * The env additions handed to `docker compose`, so a label can interpolate `${PSTACK_WILD_<P>}`.
 *
 * Compose interpolates the compose FILE from its process environment and does not re-scan the
 * substituted value, so the `$` anchor and the backticks in the rule arrive at Traefik intact — and
 * no `$$` escaping is needed, which it would be if the rule were written into the file by hand.
 */
export function subdomainEnv(routes: SubdomainRoute[]): Record<string, string> {
  const out: Record<string, string> = {};
  for (const r of routes) out[r.varName] = r.rule;
  return out;
}

File: packages/pstack/internal/spec/subdomains.go
package spec

// Wildcard subdomain routing for a per-PR surface — subdomains.ts, verbatim in intent. It lives in
// this package rather than its own because the TS split existed only to dodge an import cycle
// over SpecError; Go has no such cycle to dodge.
//
// THE GOAL. Given profile `backend` on stack `pr-123`, make anything under
// `backend-pr-123.<domain>` reach the same place the bare host reaches. pstack does not own your
// per-PR Traefik labels; it computes the one fiddly part — a correctly escaped, correctly anchored
// Go regexp — and exposes it as `PSTACK_WILD_<PROFILE>` for a label you write.
//
// THE CEILINGS: a DNS wildcard and a certificate wildcard each match exactly ONE label, so `depth:
// any` is routing-only, permanently. `one` is the default because it is what can actually be
// delivered.

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/samishal1998/preview-stacks/packages/pstack/internal/js"
	"github.com/samishal1998/preview-stacks/packages/pstack/internal/jsonx"
	"github.com/samishal1998/preview-stacks/packages/pstack/internal/omap"
)

// Depth is how many labels the wildcard spans.
type Depth string

const (
	DepthOne Depth = "one"
	DepthAny Depth = "any"
)

// SubdomainRoute is one resolved wildcard.
type SubdomainRoute struct {
	Profile string `json:"profile"`
	Host    string `json:"host"`
	Depth   Depth  `json:"depth"`
	VarName string `json:"varName"`
	Rule    string `json:"rule"`
}

// EscapeHostRegexp escapes a hostname for embedding in a Go regexp. regexp.QuoteMeta escapes the
// same fourteen bytes as the JS class `[.*+?^${}()|[\]\\]` (verified against the Go source), so the
// wake-router line and every generated HostRegexp stay byte-identical.
func EscapeHostRegexp(host string) string { return regexp.QuoteMeta(host) }

// WildcardRule is the Traefik rule for one wildcard host. Anchored at both ends; the bare host is
// deliberately NOT matched — that is the exact-host router's job.
func WildcardRule(host string, depth Depth) string {
	h := EscapeHostRegexp(host)
	const label = `[a-z0-9]([a-z0-9-]*[a-z0-9])?`
	prefix := label + `\.`
	if depth == DepthAny {
		prefix = `(` + label + `\.)+`
	}
	// Traefik v3 rules quote their argument with BACKTICKS, not quotes.
	return "HostRegexp(`^" + prefix + h + "$`)"
}

// SubdomainVarName is `backend` → `PSTACK_WILD_BACKEND`. Dashes become underscores, which makes
// `admin-api` and `admin_api` collide — the resolver checks for that.
func SubdomainVarName(profile string) string {
	return "PSTACK_WILD_" + strings.ReplaceAll(strings.ToUpper(profile), "-", "_")
}

// SubdomainConfig is one `compose.subdomains` entry before the host is resolved.
type SubdomainConfig struct {
	Profile string
	Depth   Depth
	// Host is the explicit host when the spec gave one (a pointer: `host: ""` was given, and the TS
	// `??` kept it), nil to derive `<profile>-<stack>.<domain>`.
	Host *string
}

// depthOf validates a PRESENT depth value; the callers map an absent one to DepthOne themselves,
// because a present null (`depth: ~`) is refused here with `Got null.` exactly as the TS did.
func depthOf(v any, at string) (Depth, error) {
	s, _ := v.(string)
	if s != "one" && s != "any" {
		return "", &Error{fmt.Sprintf(`%s must be "one" (a single label — what DNS and TLS can cover) or "any" (any depth, routing only: no certificate can cover it). Got %s.`, at, string(jsonx.Must(v)))}
	}
	return Depth(s), nil
}

// ParseSubdomains accepts two shapes: a list of profile names (depth one, host derived) or a mapping
// of profile → depth | { depth?, host? }. Call it only when the key is present: an absent
// `compose.subdomains` is no subdomains, but a present null is the "must be a list…" error below.
func ParseSubdomains(raw any, where string, interp func(v, at string) (string, error)) ([]SubdomainConfig, error) {
	if list, ok := raw.([]any); ok {
		out := make([]SubdomainConfig, 0, len(list))
		for i, p := range list {
			s, ok := p.(string)
			if !ok {
				return nil, &Error{fmt.Sprintf("%s[%d] must be a profile name", where, i)}
			}
			v, err := interp(s, fmt.Sprintf("%s[%d]", where, i))
			if err != nil {
				return nil, err
			}
			out = append(out, SubdomainConfig{Profile: v, Depth: DepthOne})
		}
		return out, nil
	}
	m, ok := raw.(*omap.Map)
	if !ok {
		return nil, &Error{where + " must be a list of profile names or a mapping of profile to options"}
	}
	out := make([]SubdomainConfig, 0, m.Len())
	var outer error
	m.Each(func(profile string, v any) {
		if outer != nil {
			return
		}
		at := where + "." + profile
		switch x := v.(type) {
		case nil:
			// `backend:` with no value — `v ?? undefined` in the TS, so the default depth.
			out = append(out, SubdomainConfig{Profile: profile, Depth: DepthOne})
		case string:
			d, err := depthOf(x, at)
			if err != nil {
				outer = err
				return
			}
			out = append(out, SubdomainConfig{Profile: profile, Depth: d})
		case *omap.Map:
			for _, k := range x.Keys() {
				if k != "depth" && k != "host" {
					outer = &Error{fmt.Sprintf("%s.%s is not a known option — expected `depth` or `host`", at, k)}
					return
				}
			}
			var host *string
			if hv, ok := x.Get("host"); ok {
				h, err := interp(toStr(hv), at+".host")
				if err != nil {
					outer = err
					return
				}
				host = &h
			}
			d := DepthOne
			if dv, ok := x.Get("depth"); ok {
				var err error
				if d, err = depthOf(dv, at+".depth"); err != nil {
					outer = err
					return
				}
			}
			out = append(out, SubdomainConfig{Profile: profile, Depth: d, Host: host})
		default:
			outer = &Error{fmt.Sprintf(`%s must be a depth ("one"/"any") or a mapping of { depth?, host? }`, at)}
		}
	})
	return out, outer
}

// toStr is JS String(v) over a document value: null is "null", an array joins its elements with
// "," (a null element as ""), a mapping is "[object Object]".
func toStr(v any) string {
	switch x := v.(type) {
	case nil:
		return "null"
	case []any:
		parts := make([]string, len(x))
		for i, e := range x {
			if e != nil {
				parts[i] = toStr(e)
			}
		}
		return strings.Join(parts, ",")
	case *omap.Map:
		return "[object Object]"
	default:
		return js.ToString(x)
	}
}

// ResolveSubdomains turns configs into routes. A missing domain is a hard error: a rule anchored on
// an empty suffix matches nothing, and a router that silently never fires is worse than a refusal.
func ResolveSubdomains(configs []SubdomainConfig, stack string, profiles []string, domain string) ([]SubdomainRoute, error) {
	if len(configs) == 0 {
		return []SubdomainRoute{}, nil
	}
	if domain == "" {
		return nil, &Error{"`compose.subdomains` needs a domain to anchor its rules to. Declare it in the spec (`env:\n  PREVIEW_DOMAIN: preview.example.com`) or give each entry an explicit `host:`. `DOMAIN` is accepted as a legacy alias. Without it the generated rule would match nothing."}
	}
	seen := map[string]string{}
	out := make([]SubdomainRoute, 0, len(configs))
	for _, c := range configs {
		if !contains(profiles, c.Profile) {
			list := "empty"
			if len(profiles) > 0 {
				list = strings.Join(profiles, ", ")
			}
			return nil, &Error{fmt.Sprintf(`compose.subdomains names profile "%s", which is not in compose.profiles (%s). Its router would never match a running service.`, c.Profile, list)}
		}
		varName := SubdomainVarName(c.Profile)
		if clash, ok := seen[varName]; ok {
			return nil, &Error{fmt.Sprintf(`profiles "%s" and "%s" both map to %s (a dash and an underscore are the same character in an environment variable name), so one rule would overwrite the other. Rename one profile.`, clash, c.Profile, varName)}
		}
		seen[varName] = c.Profile
		host := c.Profile + "-" + stack + "." + domain
		if c.Host != nil {
			host = *c.Host
		}
		out = append(out, SubdomainRoute{Profile: c.Profile, Host: host, Depth: c.Depth, VarName: varName, Rule: WildcardRule(host, c.Depth)})
	}
	return out, nil
}

// SubdomainEnv is the env handed to compose so a label can interpolate `${PSTACK_WILD_<P>}`.
func SubdomainEnv(routes []SubdomainRoute) map[string]string {
	out := map[string]string{}
	for _, r := range routes {
		out[r.VarName] = r.Rule
	}
	return out
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}
